sos/plugins/nvme.py

```diff
@@ -0,0 +1,119 @@
+            self.add_cmd_output("sh -c \"lsblk | grep %s | awk '{ print $4 }'\"" % dev, suggest_filename="block-size.%s" % dev)
+
+            # check if the firmware mode is OPAL
+            opal = subprocess.check_output("cat /proc/cpuinfo | grep firmware | grep OPAL | wc -l", shell=True)
+            opal = int(opal)
+            if opal == 1:
+                grep_op = "mass-storage"
+            else:
+                grep_op = "pci"
+
+            # get info about slot location and pci location
+            slot = subprocess.check_output("lscfg -vl %s | grep nvme | grep %s | awk '{ print $4 }'"
+                % (dev[0:-2], grep_op), shell=True).strip()
+            pci = subprocess.check_output("lscfg -vl %s | grep nvme | grep %s | awk '{ print $1 }'" 
+                % (dev[0:-2], grep_op), shell=True).strip()
+
+            # add slot and pci location in sos_strings/nvme/
+            self.add_string_as_file(slot, "slot_location.%s" % dev)
+            self.add_string_as_file(pci, "pci_location.%s" % dev)
+
+            # get PCI Vendor ID
+            self.add_cmd_output("sh -c \"lspci -vs %s | grep 'Non-Volatile memory controller' | cut -c 46-\"" 
+                % pci, suggest_filename="pci_vdid.%s" % dev)
+
+            # get Subsystem ID
+            self.add_cmd_output("sh -c \"lspci -vs %s | grep Subsystem | cut -c 13-\"" 
+                % pci, suggest_filename="pci_ssid.%s" % dev)
+
+            # get Part Number
+            self.add_cmd_output("sh -c \"lspci -vvs %s | grep '\[PN\]' | awk '{ print $4 }'\"" 
+                % pci, suggest_filename="part-number.%s" % dev)
+
+            # get Engineering Changes [EC] code
+            self.add_cmd_output("sh -c \"lspci -vvs %s | grep '\[EC\]' | awk '{ print $4 }'\"" 
+                % pci, suggest_filename="engineering-changes.%s" % dev)
+
+            # check if nvme-cli package is installed
+            if self.is_installed("nvme-cli"):
+                self.add_cmd_output("nvme list")
+                self.add_cmd_output("nvme list-ns /dev/%s" % dev, suggest_filename="list-ns.%s" % dev)
+                self.add_cmd_output("nvme fw-log /dev/%s" % dev, suggest_filename="fw-log.%s" % dev)
+                self.add_cmd_output("nvme list-ctrl /dev/%s" % dev, suggest_filename="list-ctrl.%s" % dev)
+                self.add_cmd_output("nvme id-ctrl -H /dev/%s" % dev, suggest_filename="id-ctrl.%s" % dev)
+                self.add_cmd_output("nvme id-ns -H /dev/%s" % dev, suggest_filename="id-ns.%s" % dev)
+                self.add_cmd_output("nvme smart-log /dev/%s" % dev, suggest_filename="smart-log.%s" % dev)
+                self.add_cmd_output("nvme error-log /dev/%s" % dev, suggest_filename="error-log.%s" % dev)
+
+            else:
+                self.add_string_as_file(("The nvme-cli tool is not installed. If you want more configuration details and" 
+                    " system information about NVMe devices, you should install it.\n"), "nvme-cli_not_installed")	
```

Comment:
You're also adding this once, for every device you discover. That's silly when it's creating a file: when it's emitting a log message it will be really annoying.